File: ring/ring.ex
# Ring benchmark
#
# Args
#   n: Number of actors
#   m: loops
#
# Usage
#
#   elixir ring.ex 2>/dev/null
#

defmodule Util do
  def debug(obj) do
    puts(obj, :stderr)
  end

  def puts(obj, out \\ :stdio) do
    IO.puts(out, "(#{inspect self()}): #{inspect obj}")
  end
end

defmodule Ring do
  def stop do
    # terminate self
    Util.debug("stopping #{inspect self()}")

    reason = :success
    exit(reason)
  end

  def process_msg(id, neighbor, count) do
    if id == 0 do
      Util.debug "master(0): #{count}"
      if count == 0 do
        stop()
      else
        new_count = count - 1
        send(neighbor, {:msg, new_count})
      end
    else
      send(neighbor, {:msg, count})
    end
  end

  def loop(id) do
    Util.debug("#{inspect self()}: start loop1")
    receive do
      {:set_neighbor, neighbor} ->
        Util.debug("#{inspect self()}: setting neighbor")
        loop(id, neighbor)
      other -> Util.debug("should not reach here: #{inspect other}")
    end
  end

  # state
  #   neighbor
  def loop(id, neighbor) do
    receive do
      {:msg, count} ->
        process_msg(id, neighbor, count)
        loop(id, neighbor)
      {:terminate} ->
        exit(:terminated_by_msg)
      other -> Util.debug("should not reach here")
    end
  end
end

# {:msg, <count>} 形式のメッセージを送受するリングループを構築する
defmodule RingLoop do
  def make_ring(m) do
    actors = (0..m) |> Enum.map(fn i -> spawn(Ring, :loop, [i]) end)
    [head|tail] = actors
    neighbors = tail ++ [head]

    Enum.zip(actors, neighbors)
      |> Enum.each(fn elem ->
        {a, n} = elem

        Util.debug("a: #{inspect a}, n: #{inspect n}")

        send(a, {:set_neighbor, n})
        end)

    actors
  end

  def send_and_receive(n, head) do
    send(head, {:msg, n})

    Util.debug("** sent")

    receive do
      {:DOWN, _ref, :process, _from_pid, reason} ->
        Util.debug({_ref, _from_pid, reason})
        reason
      {:EXIT, head, reason} ->  # 終了を補足
        Util.debug "** #{inspect head} finished (reason: #{reason})"
        reason
    after 3 * 60 * 1000 ->
      Util.debug "** timeout #{inspect head}"
      :timeout
    end
  end

  def bench_n_m(n, m) do
    ring = make_ring(n)
    [head|tail] = ring

    # Handle life of head Actor
    # https://elixirschool.com/jp/lessons/advanced/concurrency/#%E3%83%97%E3%83%AD%E3%82%BB%E3%82%B9
    _ref = Process.monitor(head)

    # time in microseconds
    {time, reason} = :timer.tc(RingLoop, :send_and_receive, [m, head])

    Enum.each(tail, fn proc -> send(proc, {:terminate}) end)

    [n, m, Float.round(time/1000/1000, 4), reason]
  end

  def run do
    # bench()
    for i <- 5..0, j <- 0..5 do
      n = round(:math.pow(10, i))
      m = round(:math.pow(10, j))

      IO.puts(Enum.join(bench_n_m(n, m), ","))
    end
  end
end

RingLoop.run

File: ring/ring.go
package main

import (
	"fmt"
	"math"
	"time"
)

type Message struct {
	msg   string
	count int
}

type Actor struct {
	id       int
	mailBox  chan Message
	neighbor *Actor
	toMain   chan Message
}

func NewActor(id int) *Actor {
	a := new(Actor)
	a.id = id
	a.mailBox = make(chan Message)
	return a
}

func (actor *Actor) ProcessMessage(msg Message) {
	if actor.id == 0 {   // master node
		msg.count -= 1
		if msg.count <= 0 {
			actor.Stop()
		}
	}

	send(actor.neighbor, msg)
}

func (actor *Actor) Stop() {
	// something
	actor.toMain <- Message{"done", -1}
}

func (actor *Actor) Loop() {
	for {
		msg := <-actor.mailBox
		actor.ProcessMessage(msg)
	}
}

func (actor *Actor) SetNeighbor(neighbor *Actor) {
	actor.neighbor = neighbor
}

func send(actor *Actor, msg Message) {
	actor.mailBox <- msg
}

func makeRing(n int) []*Actor {
	actors := make([]*Actor, n)
	for i := 0; i < n; i++ {
		actors[i] = NewActor(i)
	}

	neighbors := append([]*Actor(nil), actors...)
	neighbors = append(neighbors, neighbors[0])
	neighbors = neighbors[1:]

	for i := 0; i < n; i++ {
		actors[i].SetNeighbor(neighbors[i])
		go actors[i].Loop()
	}

	return actors
}

func benchNM(n int, m int, timeout time.Duration) {
	ring := makeRing(n)
	head := ring[0]
	head.toMain = make(chan Message)

	msg := new(Message)
	msg.count = m
	msg.msg = "hey"

	var reason string
	start := time.Now()
	send(head, *msg)
	select {
	case <-head.toMain:
		reason = "succeeded"
	case <-time.After(timeout):
		reason = "timeout"
	}
	elapsed := time.Since(start)
	fmt.Printf("%d,%d,%.4f,%s\n", n, m, elapsed.Seconds(), reason)
}

func main() {
	var n, m int
	timeout := time.Second * 3 * 60

	fmt.Println("n,m,elapsed(s),reason")

	for i := 1; i < 6; i++ {
		for j := 1; j < 6; j++ {
			n = int(math.Pow10(i))
			m = int(math.Pow10(j))
			benchNM(n, m, timeout)
		}
	}
}
